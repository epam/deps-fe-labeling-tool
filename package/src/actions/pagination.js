import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
