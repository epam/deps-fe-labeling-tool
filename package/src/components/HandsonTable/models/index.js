
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
