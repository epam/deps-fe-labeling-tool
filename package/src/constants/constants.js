const CONFIDENCE_ON_MANUAL_CHANGE = 1

export {
  CONFIDENCE_ON_MANUAL_CHANGE
}
