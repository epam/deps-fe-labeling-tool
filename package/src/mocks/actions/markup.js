const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
