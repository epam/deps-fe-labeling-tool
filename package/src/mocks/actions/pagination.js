const openPage = jest.fn((page) => ({
  type: 'MOCK_OPEN_PAGE_ACTION',
  payload: page
}))

const mockPaginationActions = {
  openPage
}

export {
  mockPaginationActions
}
