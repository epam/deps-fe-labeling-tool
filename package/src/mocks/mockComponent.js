const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
