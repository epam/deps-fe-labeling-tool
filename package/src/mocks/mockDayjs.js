
const formats = {
  L: 'MM/DD/YYYY',
  LT: 'h:mm A',
  default: 'MM/DD/YYYY, HH:mm'
}

const MOCKED_DATE = '2023-10-03T12:14:02'

const mockDayjs = (dateTimeStr = MOCKED_DATE) => ({
  __esModule: true,
  default: (() => {
    function dayjs (date = dateTimeStr) {
      return {
        valueOf: () => new Date(date).getTime(),
        format: () => date,
        isValid: () => true,
        diff: () => -100500,
        add: () => dayjs(date),
        startOf: () => dayjs(date),
        endOf: () => dayjs(date),
        subtract: () => dayjs(date),
        longDateFormat: (format = 'default') => formats[format],
        localeData: () => dayjs(date),
        toISOString: () => date
      }
    }
    dayjs.locale = () => {}
    dayjs.extend = () => {}
    dayjs.utc = () => dayjs(dateTimeStr)
    return dayjs
  })()
})

export {
  mockDayjs
}
