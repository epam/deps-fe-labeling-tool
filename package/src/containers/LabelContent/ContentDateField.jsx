import React, { useCallback } from 'react'
import dayjs from 'dayjs'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateLabels } from '@/actions/markup'
import { CalendarIcon } from '@/components/Icons/CalendarIcon'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { debounce } from '@/hocs/debounce'
import { labelShape } from '@/models/Label'
import { currentPageSelector } from '@/selectors/pagination'
import {
  LOCALE_DATE_FORMAT,
  dayjsToString,
  stringToDayjs
} from '@/utils/dayjs'
import {
  InputWrapper,
  StyledDatePicker,
  StyledInput
} from './ContentDateField.styles'

const DebouncedInput = debounce({
  getValue: (e) => e.target.value
})(StyledInput)

const dateNow = dayjs()

const ContentDateField = ({
  currentPage,
  label,
  updateLabels
}) => {
  const handleLabelsUpdate = useCallback(
    (content) => {
      updateLabels(currentPage, [{
        ...label,
        content,
        confidence: CONFIDENCE_ON_MANUAL_CHANGE
      }])
    },
    [
      currentPage,
      label,
      updateLabels
    ])

  const handleDatePick = useCallback(
    (dayjsInstance) =>
      handleLabelsUpdate(dayjsToString(dayjsInstance, LOCALE_DATE_FORMAT)),
    [handleLabelsUpdate]
  )

  return (
    <>
      <InputWrapper>
        <DebouncedInput
          value={label.content}
          onChange={handleLabelsUpdate}
        />
        <StyledDatePicker
          allowClear={false}
          value={
            stringToDayjs(label.content) ||
            dateNow
          }
          onChange={handleDatePick}
          suffixIcon={<CalendarIcon />}
        />
      </InputWrapper>
    </>
  )
}

ContentDateField.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state)
})

const mapDispatchToProps = {
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ContentDateField)

export {
  ConnectedComponent as ContentDateField
}
