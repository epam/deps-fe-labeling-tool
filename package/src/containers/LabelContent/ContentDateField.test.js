import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockComponent } from '@/mocks/mockComponent'
import { mockDayjs } from '@/mocks/mockDayjs'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import dayjs from 'dayjs'
import { shallow } from 'enzyme'
import { updateLabels } from '@/actions/markup'
import { CONFIDENCE_ON_MANUAL_CHANGE } from '@/constants/constants'
import { Label, LabelType } from '@/models/Label'
import { dayjsToString } from '@/utils/dayjs'
import { ContentDateField } from './ContentDateField'
import { StyledDatePicker, StyledInput } from './ContentDateField.styles'

jest.mock('dayjs', () => mockDayjs())
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/containers/SpecialSymbols', () =>
  mockComponent('SpecialSymbols')
)
jest.mock('@/hocs/debounce', () => ({
  debounce: () => (Component) => Component
}))

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = ContentDateField

describe('Container: ContentDateField', () => {
  describe('mapStateToProps', () => {
    it('should pass expected state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      const expectedState = {
        currentPage: mockPaginationSelectors.currentPageSelector()
      }
      expect(props).toEqual(expectedState)
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass updateLabels action as prop to WrappedComponent', () => {
      props.updateLabels()
      expect(updateLabels).toHaveBeenCalled()
    })
  })

  describe('Component', () => {
    let wrapper, defaultProps

    beforeEach(() => {
      defaultProps = {
        ...mapStateToProps().props,
        ...mapDispatchToProps().props,
        label: new Label(
          0.113,
          0.8104,
          0.017,
          0.0119,
          'enclosures',
          undefined,
          LabelType.CHECKMARK,
          'date',
          { data: 'some meta for enclosures value label' },
          0.66
        )
      }
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout based on props', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call updateLabels prop with correct args when calling StyledInput onChange', () => {
      const TextAreaWrapper = wrapper.find(StyledInput)
      const mockValue = 'mockValue'
      TextAreaWrapper.props().onChange(mockValue)

      const expectedSecondArg = [
        {
          ...defaultProps.label,
          content: mockValue,
          confidence: CONFIDENCE_ON_MANUAL_CHANGE
        }
      ]

      expect(updateLabels).nthCalledWith(
        1,
        defaultProps.currentPage,
        expectedSecondArg
      )
    })

    it('should call updateLabels prop with correct args when calling StyledDatePicker onChange', () => {
      const TextAreaWrapper = wrapper.find(StyledDatePicker)
      const mockValue = dayjs()
      TextAreaWrapper.props().onChange(mockValue)

      const expectedSecondArg = [
        {
          ...defaultProps.label,
          content: dayjsToString(mockValue),
          confidence: CONFIDENCE_ON_MANUAL_CHANGE
        }
      ]

      expect(updateLabels).nthCalledWith(
        1,
        defaultProps.currentPage,
        expectedSecondArg
      )
    })
  })
})
