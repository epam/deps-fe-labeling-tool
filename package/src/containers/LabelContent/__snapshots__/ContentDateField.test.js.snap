// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ContentDateField Component should render correct layout based on props 1`] = `
<Fragment>
  <ContentDateFieldstyles__InputWrapper>
    <ContentDateFieldstyles__StyledInput
      onChange={[Function]}
      value="date"
    />
    <ContentDateFieldstyles__StyledDatePicker
      allowClear={false}
      onChange={[Function]}
      suffixIcon={<CalendarIcon />}
      value={
        Object {
          "add": [Function],
          "diff": [Function],
          "endOf": [Function],
          "format": [Function],
          "isValid": [Function],
          "localeData": [Function],
          "longDateFormat": [Function],
          "startOf": [Function],
          "subtract": [Function],
          "toISOString": [Function],
          "valueOf": [Function],
        }
      }
    />
  </ContentDateFieldstyles__InputWrapper>
</Fragment>
`;
